File: generated_code/cpp/solution_3774_1.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    bool hasSameDigits(string s) {
        int n = s.length();
        int num1 = 0;
        int num2 = 0;
        
        // This is equivalent to finding the contribution of each element to the 
        // final two digits. The transformation follows Pascal's triangle.
        // For a string of length n, the final two elements are sums of s[i] * C(n-2, i) mod 10.
        // We compute C(n-2, i) mod 10 using Lucas' Theorem for primes 2 and 5, then combine using CRT.
        for (int i = 0; i <= n - 2; ++i) {
            int coefficient = nCkMod10(n - 2, i);
            
            num1 = (num1 + coefficient * (s[i] - '0')) % 10;
            num2 = (num2 + coefficient * (s[i + 1] - '0')) % 10;
        }
        
        return num1 == num2;
    }

private:
    // Returns nCr % 10 using the Chinese Remainder Theorem (CRT) with Lucas' Theorem.
    int nCkMod10(int n, int k) {
        int mod2 = lucasTheorem(n, k, 2);
        int mod5 = lucasTheorem(n, k, 5);
        
        // Lookup table combining mod 2 and mod 5 answers via CRT
        // lookup[mod2][mod5]
        static constexpr int lookup[2][5] = {
            {0, 6, 2, 8, 4}, // mod2 == 0
            {5, 1, 7, 3, 9}  // mod2 == 1
        };
        
        return lookup[mod2][mod5];
    }

    // Lucas' Theorem computes (nCr) % prime in O(log_prime(n)) time
    int lucasTheorem(int n, int k, int prime) {
        int res = 1;
        while (n > 0 || k > 0) {
            int nMod = n % prime;
            int kMod = k % prime;
            
            if (kMod > nMod) return 0; // nCr is 0 if r > n
            
            res = (res * nCk(nMod, kMod)) % prime;
            
            n /= prime;
            k /= prime;
        }
        return res;
    }

    // Simple nCr for small values (since arguments are bounded by the prime base)
    int nCk(int n, int k) {
        if (k > n) return 0;
        if (k == 0 || k == n) return 1;
        
        int res = 1;
        for (int i = 0; i < k; ++i) {
            res *= (n - i);
            res /= (i + 1);
        }
        return res;
    }
};

File: generated_code/python/solution_3774_1.py
from typing import List

class Solution:
    def hasSameDigits(self, s: str) -> bool:
        """
        The process of repeatedly taking adjacent sums modulo 10 is equivalent to 
        finding the coefficients of the digits using Pascal's triangle (combinations).
        For a string of length n, reducing it down to 2 digits means we perform the 
        operation n - 2 times.
        
        The first final digit (left) is a combination of the first n-1 digits:
        left = sum( C(n-2, i) * s[i] ) % 10 for i from 0 to n-2
        
        The second final digit (right) is a combination of the last n-1 digits:
        right = sum( C(n-2, i) * s[i+1] ) % 10 for i from 0 to n-2
        
        Since we need the results modulo 10, and 10 is not prime (10 = 2 * 5), 
        we can compute the binomial coefficients modulo 2 and modulo 5 separately 
        using Lucas' Theorem, and then combine them using the Chinese Remainder Theorem (CRT).
        
        Time Complexity: O(n) to iterate through the string and compute the combinations.
        Space Complexity: O(1) auxiliary space.
        """
        n = len(s)
        m = n - 2
        
        # Precompute small combinations for Lucas' Theorem
        # nCr % 2 table
        c2 = [[0] * 2 for _ in range(2)]
        for i in range(2):
            c2[i][0] = 1
            for j in range(1, i + 1):
                c2[i][j] = (c2[i-1][j-1] + c2[i-1][j]) % 2
                
        # nCr % 5 table
        c5 = [[0] * 5 for _ in range(5)]
        for i in range(5):
            c5[i][0] = 1
            for j in range(1, i + 1):
                c5[i][j] = (c5[i-1][j-1] + c5[i-1][j]) % 5

        def nCr_mod_p(n: int, r: int, p: int, table: List[List[int]]) -> int:
            # Lucas' Theorem: nCr % p = Product( n_i C r_i % p )
            ans = 1
            while n > 0 or r > 0:
                ni = n % p
                ri = r % p
                if ri > ni:
                    return 0
                ans = (ans * table[ni][ri]) % p
                n //= p
                r //= p
            return ans

        # Chinese Remainder Theorem for combining mod 2 and mod 5 to mod 10
        # x % 2 = r2, x % 5 = r5 -> x % 10
        # By inspection: x = (r2 * 5 * 1 + r5 * 2 * 3) % 10 = (5 * r2 + 6 * r5) % 10
        def crt_2_5(r2: int, r5: int) -> int:
            return (5 * r2 + 6 * r5) % 10

        left_sum = 0
        right_sum = 0
        
        # Calculate the contributions of each digit
        for i in range(m + 1):
            r2 = nCr_mod_p(m, i, 2, c2)
            r5 = nCr_mod_p(m, i, 5, c5)
            coeff = crt_2_5(r2, r5)
            
            left_sum = (left_sum + coeff * int(s[i])) % 10
            right_sum = (right_sum + coeff * int(s[i+1])) % 10
            
        return left_sum == right_sum